examples/cevae.py

```diff
@@ -0,0 +1,311 @@
+    https://github.com/AMLab-Amsterdam/CEVAE
+"""
+import argparse
+
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader, TensorDataset
+
+import pyro
+import pyro.distributions as dist
+from pyro import poutine
+from pyro.infer import SVI, Trace_ELBO
+from pyro.infer.util import torch_item
+from pyro.nn import PyroModule
+from pyro.optim import ClippedAdam
+
+
+class FullyConnected(nn.Sequential):
+    def __init__(self, sizes):
+        layers = []
+        for in_size, out_size in zip(sizes, sizes[1:]):
+            layers.append(nn.Linear(in_size, out_size))
+            layers.append(nn.ELU())
+        layers.pop(-1)
+        super().__init__(*layers)
+
+
+def nn_init(m):
+    if isinstance(m, nn.Linear):
+        m.weight.data.normal_(0, 0.1 / m.in_features ** 0.5)
+        m.bias.data.fill_(0)
+
+
+class DiagNormalNet(nn.Module):
+    """
+    FullyConnected network outputting a constrained ``loc,scale`` pair.
+    """
+    def __init__(self, sizes):
+        super().__init__()
+        self.fc = FullyConnected(sizes[:-1] + [sizes[-1] * 2])
+
+    def forward(self, x):
+        loc_scale = self.fc(x)
+        d = loc_scale.size(-1) // 2
+        loc = loc_scale[..., :d]
+        scale = nn.functional.softplus(loc_scale[..., d:]).clamp(min=1e-10)
+        return loc, scale
+
+
+class BernoulliNet(nn.Module):
+    """
+    FullyConnected network outputting a single ``logits`` value.
+    """
+    def __init__(self, sizes):
+        super().__init__()
+        self.fc = FullyConnected(sizes + [1])
+
+    def forward(self, x):
+        return self.fc(x).squeeze(-1)
+
+
+class Model(PyroModule):
+    """
+    Generative model for a causal model with latent confounder::
+
+        z ~ p(z)      # latent confounder
+        x ~ p(x|z)    # partial noisy observation of z
+        t ~ p(t|z)    # treatment, whose application is biased by z
+        y ~ p(y|t,z)  # outcome
+
+    Each of these distributions is defined by a neural network.  The ``y``
+    distribution is defined by a disjoint pair of neural networks defining
+    ``p(y|t=0,z)`` and ``p(y|t=1,z)``; this allows highly imbalanced treatment.
+    """
+    def __init__(self, args):
+        self.latent_dim = args.latent_dim
+        super().__init__()
+        self.x_nn = DiagNormalNet([args.latent_dim] +
+                                  [args.hidden_dim] * args.num_layers +
+                                  [args.feature_dim])
+        self.y0_nn = BernoulliNet([args.latent_dim] +
+                                  [args.hidden_dim] * args.num_layers)
+        self.y1_nn = BernoulliNet([args.latent_dim] +
+                                  [args.hidden_dim] * args.num_layers)
+        self.t_nn = BernoulliNet([args.latent_dim])
+
+    def forward(self, x, t=None, y=None, size=None):
+        if size is None:
+            size = x.size(0)
+        with pyro.plate("data", size, subsample=x):
```

Comment:
nit: use `subsample_size=x.size(0)` explicitly here for clarity since `x` isn't actually an index